core/dbt/task/clone.py

```diff
@@ -0,0 +1,183 @@
+        for node in self.manifest.nodes.values():
+            if node.unique_id not in selected_uids:
+                continue
+            if node.is_relational and not node.is_ephemeral:
+                relation = adapter.Relation.create_from(self.config, node)
+                result.add(relation.without_identifier())
+
+                # cache the 'other' schemas too!
+                if node.state_relation:  # type: ignore
+                    other_relation = adapter.Relation.create_from_node(
+                        self.config, node.state_relation  # type: ignore
+                    )
+                    result.add(other_relation.without_identifier())
+
+        return result
+
+    def before_run(self, adapter, selected_uids: AbstractSet[str]):
+        with adapter.connection_named("master"):
+            # unlike in other tasks, we want to add information from the --state manifest *before* caching!
+            self.defer_to_manifest(adapter, selected_uids)
+            # only create *our* schemas, but cache *other* schemas in addition
+            schemas_to_create = super().get_model_schemas(adapter, selected_uids)
+            self.create_schemas(adapter, schemas_to_create)
+            schemas_to_cache = self.get_model_schemas(adapter, selected_uids)
+            self.populate_adapter_cache(adapter, schemas_to_cache)
+
+    @property
+    def resource_types(self):
+        if not self.args.resource_types:
+            return NodeType.refable()
+
+        values = set(self.args.resource_types)
+
+        if "all" in values:
+            values.remove("all")
+            values.update(NodeType.refable())
+
+        values = [NodeType(val) for val in values if val in NodeType.refable()]
+
+        return list(values)
+
+    def get_node_selector(self) -> ResourceTypeSelector:
+        resource_types = self.resource_types
+
+        if self.manifest is None or self.graph is None:
+            raise DbtInternalError("manifest and graph must be set to get perform node selection")
+        return ResourceTypeSelector(
+            graph=self.graph,
+            manifest=self.manifest,
+            previous_state=self.previous_state,
+            resource_types=resource_types,
+        )
+
+    def get_runner_type(self, _):
+        return CloneRunner
+
+    def _get_deferred_manifest(self) -> Optional[WritableManifest]:
+        state = self.previous_state
+        if state is None:
+            raise DbtRuntimeError(
+                "--state is required for cloning relations from another environment"
+            )
+
+        if state.manifest is None:
+            raise DbtRuntimeError(f'Could not find manifest in --state path: "{self.args.state}"')
+        return state.manifest
+
+    # Note that this is different behavior from --defer with other commands, which *merge*
+    # selected nodes from this manifest + unselected nodes from the other manifest
+    def defer_to_manifest(self, adapter, selected_uids: AbstractSet[str]):
```

Comment:
It's compelling to want access to `state_relation` much more frequently. If we have it, we can do cool things with it; `clone` is just one of those cool things. This has come up in a handful of conversations over the past few weeks/months:
- Inferring model contracts/diffs based on the prod db object (#7432)
- Interactive preview + `audit_helper` to calculate diffs between prod/dev

I know that was split off as a separate bit of work, and it was already implemented (#7796). It might be worth reconsidering this as either:
- not mutually exclusive with `merge_from_artifact`
- the default behavior when `--defer` is set, rather than something totally unique and special to the `clone` command

My intent in this comment is to share context - no action needed right now.

Reply:
I totally agree with removing `state_relation` before writing the manifest. That's exactly how I (naively) implemented this in my spike PR: https://github.com/dbt-labs/dbt-core/pull/7258/files#r1155323650

Reply:
Thinking more about this: Do we actually need separate methods for `merge_from_artifact` and `add_from_artifact`? Or could it simply be that:
- for nodes which are "being deferred," we completely clobber/replace this manifest's version with the other manifest's version
- for nodes which _aren't_ "being deferred" (because they're being selected, or already exist in this schema), we keep the existing node entry and add the `defer_relation` attribute

Then, all tasks could call this same method. We don't need the divergent behavior between `clone` and `run`, and we set ourselves up more nicely for future work around contract inference, dev/prod diff, ...

Pseudo code:
```python
    def merge_from_artifact(
        self,
        adapter,
        other: "WritableManifest",
        selected: AbstractSet[UniqueID],
        favor_state: bool = False,
    ) -> None:
        """Given the selected unique IDs and a writable manifest, update this
        manifest by replacing any unselected nonexisting nodes with their counterpart.
        For any selected/existing nodes, add information about each node's location
        in the other manifest.

        Only non-ephemeral refable nodes are examined.
        """
        refables = set(NodeType.refable())
        merged = set()
        for unique_id, node in other.nodes.items():
            current = self.nodes.get(unique_id)
            
            # do the replace thing
            if current and (
                node.resource_type in refables
                and not node.is_ephemeral
                and unique_id not in selected
                and (
                    not adapter.get_relation(current.database, current.schema, current.identifier)
                    or favor_state
                )
            ):
                merged.add(unique_id)
                self.nodes[unique_id] = node.replace(deferred=True)
           
           # do the additive thing
           elif current and (node.resource_type in refables and not node.is_ephemeral):
                defer_state_relation = RelationalNode(node.database, node.schema, node.alias)
                self.nodes[unique_id] = current.replace(defer_relation=defer_relation)

        # Rebuild the flat_graph, which powers the 'graph' context variable,
        # now that we've deferred some nodes
        self.build_flat_graph()

        # log up to 5 items
        sample = list(islice(merged, 5))
        fire_event(MergedFromState(num_merged=len(merged), sample=sample))
```

The complication I could foresee here: Traditional defer behavior requires the population of the adapter cache (for the "nonexisting" part). I had suggested [doing things differently in `clone`'s `before_run` setup](https://github.com/dbt-labs/dbt-core/blob/4bce05fb0c991d239a9d9d8a5a392dee5d3502d2/core/dbt/task/clone.py#L118-L126) so that we defer _first_, and then cache, so that we can also cache the ["other" schemas (indicated by `defer_relation`)](https://github.com/dbt-labs/dbt-core/blob/4bce05fb0c991d239a9d9d8a5a392dee5d3502d2/core/dbt/task/clone.py#L109-L114). I'm not sure if that's actually necessary, or if it's extra complication that isn't really worth it.